src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,656 @@
+        return methodSymbol?.IsStatic is true;
+    }
+
+    /// <summary>
+    /// Checks whether the constructor parameter is in use in the expression.
+    /// </summary>
+    /// <param name="expression">The expression to evaluate.</param>
+    /// <param name="context">The analysis context.</param>
+    /// <returns>A bool, on whether we've found a constructor parameter in use.</returns>
+    private static bool IsConstructorParameterInUse(ExpressionSyntax expression, SyntaxNodeAnalysisContext context)
+    {
+        return expression switch
+        {
+            // There are many ways of initializing the same field, so we need to check many different types of expressions.
+            // For example, new List<string>() == [] == new List<string> { }. So we need to consider these scenarios.
+            InvocationExpressionSyntax invocationExpressionSyntax => IsConstructorParameterInUse(invocationExpressionSyntax.ArgumentList, context),
+            ObjectCreationExpressionSyntax objectCreationExpressionSyntax => IsConstructorParameterInUse(objectCreationExpressionSyntax.ArgumentList, context),
+            ImplicitObjectCreationExpressionSyntax implicitObjectCreationExpressionSyntax => IsConstructorParameterInUse(implicitObjectCreationExpressionSyntax.ArgumentList, context),
+            InitializerExpressionSyntax initializerExpressionSyntax => IsConstructorParameterInUse(initializerExpressionSyntax.Expressions, context),
+            _ => false,
+        };
+    }
+
+    // Overload for SyntaxNode
+    private static bool IsConstructorParameterInUse(SyntaxNode? argumentList, SyntaxNodeAnalysisContext context)
+    {
+        return argumentList is not null && AreIdentifierNodesReferencingConstructorParameters(argumentList.DescendantNodes().OfType<IdentifierNameSyntax>(), context);
+    }
+
+    // Overload for SeparatedSyntaxList<ExpressionSyntax>
+    private static bool IsConstructorParameterInUse(SeparatedSyntaxList<ExpressionSyntax> expressions, SyntaxNodeAnalysisContext context)
+    {
+        return AreIdentifierNodesReferencingConstructorParameters(expressions.OfType<IdentifierNameSyntax>(), context);
+    }
+
+    /// <summary>
+    /// Checks whether the constructor parameter is in use in the given nodes.
+    /// </summary>
+    /// <param name="identifiers">The <see cref="IEnumerable{IdentifierNameSyntax}"/> to check for constructor parameter usage.</param>
+    /// <param name="context">The analysis context.</param>
+    /// <returns>A bool, on whether a constructor parameter was found in the nodes to check.</returns>
+    private static bool AreIdentifierNodesReferencingConstructorParameters(IEnumerable<IdentifierNameSyntax> identifiers, SyntaxNodeAnalysisContext context)
+    {
+        foreach (IdentifierNameSyntax identifierNameSyntax in identifiers)
+        {
+            IOperation? operation = context.SemanticModel.GetOperation(identifierNameSyntax, context.CancellationToken);
```

Comment:
Note: this is expensive

Reply:
I'm changing it so we just assume an identifier being used in a constructor with a parameter means the expression is dependent on the constructor. This is because it is too expensive to rely on the semantic model and given the amount of edge cases, it is much safer.